### DISCUSSION TITLE
System set with run criteria ignoring AppState

### DISCUSSION THREAD

**Discussion by Jeremy-Code-F** [bevyengine/bevy#7413] [Q&A] [answered]
Hi,

I've been following https://bevy-cheatbook.github.io/programming/states.html to implement app states in my snake clone game.

For some reason it seems like bevy ignores the AppState when using ```with_run_criteria```

```
.add_system_set(
            SystemSet::on_update(AppState::InGame)
                .with_run_criteria(FixedTimestep::step(1.0))
                .with_system(food_spawner),
```

I would expect the above code to only call ```food_spawner``` when the AppState is in ```AppState::InGame``` but the function just gets called anyways.

If I change the call to ```add_system_set``` to...

```
.add_system_set(
            SystemSet::on_update(AppState::InGame)
                .with_system(food_spawner),
```

Then it works the way I would expect, but of course I lose my FixedTimestep.

Am I doing something wrong here?

p.s. similar question [here](https://github.com/bevyengine/bevy/discussions/2763) but there was no answer on that post.

**Comment by rparrett** [comment, marked as answer]
You're not doing anything wrong. This is a limitation in Bevy 0.9.

The text on the cheatbook page you linked to hints at this:
> Consider using the [iyes_loopless](https://github.com/IyesGames/iyes_loopless) crate, which provides an alternative implementation that does not suffer from the [usability issues](https://bevy-cheatbook.github.io/programming/states.html#known-pitfalls-and-limitations) of the one in Bevy.

And following the `usability issue` link explains further:
> Because states are implemented using [run criteria](https://bevy-cheatbook.github.io/programming/run-criteria.html), they cannot be combined with other uses of run criteria, such as [fixed timestep](https://bevy-cheatbook.github.io/features/fixed-timestep.html).

This limitation is being addressed by an ongoing scheduler overhaul detailed in https://github.com/bevyengine/rfcs/pull/45 which will likely be making it into the next release of Bevy.
